Implement basic Ledger class which creates Blocks

AndrejMitrovic opened the pull request:
This is a draft PR. Only the last commit is relevant. Requires https://github.com/bpfkorea/agora/pull/66

I'll a wait comments from @Geod24. In particular I'm wondering where to put the Ledger instance.. Perhaps just in the Node? But then the GossipProtocol will have to have an instance of the Node. 

It would be good to get some design ideas.

Geod24 commented on source/agora/node/GossipProtocol.d:
```diff
@@ -61,6 +66,7 @@ public class GossipProtocol
 
         this.tx_cache[tx_hash] = tx;
         this.network.sendTransaction(tx);
+        this.ledger.receiveTransaction(tx);
```
Local storage before network communication

AndrejMitrovic replied:
Yeah I was wondering about that actually. Could we add it to the style guide?

Geod24 commented on source/agora/common/Block.d:
```diff
@@ -55,3 +56,51 @@ public struct Block
+{
+    Block block;
+
+    block.header.height = 0;  // genesis block is 0
```
That's the default?

Geod24 replied:
However, there should be a transaction giving 40k to a `random()` keypair.

AndrejMitrovic replied:
Makes sense.

AndrejMitrovic replied:
Speaking of which, we need to define the unitary model or just a helper function (so 1 BOA = 1,000,000 units or whatever we defined it as). So I could do something like `output.amount = 1.boa` and `boa` is defined as `long boa (long amount) { return amount * 1_000_000; }`.

That could be a separate task in the backlog.

Geod24 replied:
Correct

Geod24 commented:
Looks pretty neat

AndrejMitrovic committed:
Add helper functions for creating blocks

getGenesisBlock() returns the genesis block,
which will later be refined.

makeNewBlock() helps create a new block,
referencing the header of a previous block.

AndrejMitrovic committed:
Implement the block tracking Ledger class

It contains a simple API for receiving transactions
and internally creating new blocks and adding them
to the internal ledger.

AndrejMitrovic commented:
Rebased and reworked it a little bit.

The unittests could perhaps use some refinement.

Geod24 commented on source/agora/common/Block.d:
```diff
@@ -95,3 +95,78 @@ public struct Block
+/*******************************************************************************
+
+    Creates the genesis block.
+    Note: The output specifies a randomly-generated address,
```
```suggestion

    The output specifies a randomly-generated address,
```

Otherwise you create a DDOC section

AndrejMitrovic replied:
That node is outdated anyway. :D

Geod24 commented on source/agora/common/Block.d:
```diff
@@ -95,3 +95,78 @@ public struct Block
+
+*******************************************************************************/
+
+public Block getGenesisBlock ( )
```
```suggestion
public Block getGenesisBlock ()
```

AndrejMitrovic replied:
Ah, makes it easier to grep? Add it to the style guide!

Geod24 replied:
It already is there: https://github.com/bpfkorea/agora/pull/74/files#diff-f69cd4b5588d09bb541f85596fdacad3R57

Geod24 commented on source/agora/common/Block.d:
```diff
@@ -95,3 +95,78 @@ public struct Block
+
+    // note: not using random() as the genesis block should always be the same
+    auto address = KeyPair.fromSeed(
+        Seed.fromString("SCT4KKJNYLTQO4TVDPVJQZEONTVVW66YLRWAINWI3FZDY7U4JS4JJEI4")).address;
```
I should make this CTFE-able so we can just use `static immutable`

AndrejMitrovic replied:
yessss please.

Geod24 commented on source/agora/common/Block.d:
```diff
@@ -95,3 +95,78 @@ public struct Block
+        prev_block = the previous block
+        tx = the transaction that will be contained in the new block
+
+    In the current preliminary design a block contains a single transaction.
```
Move this before `Params`

Geod24 commented on source/agora/common/Block.d:
```diff
@@ -95,3 +95,78 @@ public struct Block
+
+*******************************************************************************/
+
+public Block makeNewBlock ( Block prev_block, Transaction tx ) @nogc nothrow @safe
```
```suggestion
public Block makeNewBlock (Block prev_block, Transaction tx) @nogc nothrow @safe
```

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -0,0 +1,99 @@
+import agora.common.Transaction;
+
+/// Ditto
+class Ledger
```
```suggestion
public class Ledger
```

Geod24 replied:
Pin g

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -0,0 +1,99 @@
+    private Block[Hash] ledger;
+
+    /// latest block height
+    private Block last_block;
```
Pointer ?

AndrejMitrovic replied:
That's an extra dereference I guess, but we're fine with it?

Geod24 replied:
I take a dereference over a copy any time of the day.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -0,0 +1,99 @@
+
+    ***************************************************************************/
+
+    public Block getLastBlock ( ) @safe nothrow @nogc
```
```suggestion
    public Block getLastBlock () @safe nothrow @nogc
```

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -0,0 +1,99 @@
+
+    ***************************************************************************/
+
+    private void addNewBlock ( Block block ) @safe nothrow
```
```suggestion
    private void addNewBlock (Block block) @safe nothrow
```

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -0,0 +1,99 @@
+    auto rand_pair = KeyPair.random();
+    auto input = Input(hashFull(42));
+    input.signature = rand_pair.secret.sign(input.hashFull[]);
+    Transaction tx = { inputs : [input], outputs: [Output(100, rand_pair.address)]};
```
```suggestion
    Transaction tx = { inputs : [input], outputs: [ Output(100, rand_pair.address) ] };
```

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -0,0 +1,99 @@
+
+    ***************************************************************************/
+
+    public void receiveTransaction (Transaction tx) @trusted
```
That's very network-oriented. What about loading from disk ? Will the name stick ?

AndrejMitrovic replied:
Hmm I mean I could name it `addTransaction`, but I was worried about this name because the transaction might be invalid, therefore it will **not** be added, despite the name.

So what do we call it?

Geod24 replied:
Damn hard questions. We can go with this now and change later.

AndrejMitrovic commented:
Updated PR with the suggested changes.

AndrejMitrovic commented:
Missed out the public thing. Done.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -0,0 +1,101 @@
+    {
+        auto block_hash = block.header.hashFull();
+        this.ledger[block_hash] = block;
+        this.last_block = block_hash in this.ledger;
```
This is a very crude approach but will work for now :)
